pkg/openstackbackup/funcs.go

```diff
@@ -0,0 +1,518 @@
+	crLists := ospdirectorv1beta1.CrsForBackup{}
+
+	listOpts := []client.ListOption{
+		client.InNamespace(namespace),
+	}
+
+	// OpenStackBaremetalSets
+
+	osBms := ospdirectorv1beta1.OpenStackBaremetalSetList{}
+
+	if err := r.GetClient().List(context.Background(), &osBms, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackBaremetalSets = osBms
+
+	// OpenStackClients
+
+	osClients := ospdirectorv1beta1.OpenStackClientList{}
+
+	if err := r.GetClient().List(context.Background(), &osClients, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackClients = osClients
+
+	// OpenStackControlPlanes
+
+	osCtlPlanes := ospdirectorv1beta1.OpenStackControlPlaneList{}
+
+	if err := r.GetClient().List(context.Background(), &osCtlPlanes, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackControlPlanes = osCtlPlanes
+
+	// OpenStackIPSets
+
+	osIps := ospdirectorv1beta1.OpenStackIPSetList{}
+
+	if err := r.GetClient().List(context.Background(), &osIps, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackIPSets = osIps
+
+	// OpenStackMACAddresses
+
+	osMacAddresses := ospdirectorv1beta1.OpenStackMACAddressList{}
+
+	if err := r.GetClient().List(context.Background(), &osMacAddresses, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackMACAddresses = osMacAddresses
+
+	// OpenStackNetworks
+
+	osNets := ospdirectorv1beta1.OpenStackNetList{}
+
+	if err := r.GetClient().List(context.Background(), &osNets, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackNets = osNets
+
+	// OpenStackProvisionServers
+
+	osProvServers := ospdirectorv1beta1.OpenStackProvisionServerList{}
+
+	if err := r.GetClient().List(context.Background(), &osProvServers, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackProvisionServers = osProvServers
+
+	// OpenStackVMSets
+
+	osVms := ospdirectorv1beta1.OpenStackVMSetList{}
+
+	if err := r.GetClient().List(context.Background(), &osVms, listOpts...); err != nil {
+		return crLists, err
+	}
+
+	crLists.OpenStackVMSets = osVms
+
+	return crLists, nil
+}
+
+// GetConfigMapList - Get list of all OSP-D operator config maps in the namespace that we care to save/restore
+func GetConfigMapList(r common.ReconcilerCommon, request *ospdirectorv1beta1.OpenStackBackupRequest, desiredCrs *ospdirectorv1beta1.CrsForBackup) (corev1.ConfigMapList, error) {
+	configMapList := &corev1.ConfigMapList{}
+
+	labels := client.HasLabels{
+		common.OwnerControllerNameLabelSelector,
+	}
+
+	listOpts := []client.ListOption{
+		client.InNamespace(request.Namespace),
+		labels,
+	}
+
+	err := r.GetClient().List(context.TODO(), configMapList, listOpts...)
+
+	if err != nil {
+		return *configMapList, err
+	}
+
+	// Also get certain config maps in other particular namespaces?
+	// TODO: which ones?
+
+	// Also get certain config maps used by our CRs
+	// TODO: which ones?
```

Comment:
we should backup the CMs/Secretes added via parameter to all CRs, like osplaybookgenerator: tarballConfigMap, gitSecret, heatEnvConfigMap

Reply:
or the controllers add backup label to the CM/Secret passed via a parameter so this can be used to filter instead of a named list

Reply:
We could look for existing `OsPlaybookGenerators` during backup and copy the CMs and Secrets.  I initially left them out since we do not save `OsPlaybookGenerators`.  I believe CMs and Secrets included as params in `OsPlaybookGenerators` don't have any ownership set on them, so I think it would be fine to save them without having the `OsPlaybookGenerators` in the backup.

To your second point: yes, we could also slap the controller label that I already look for on the CMs and Secrets when we processes `OsPlaybookGenerators`.

Reply:
I just thought it would be good to save the inputs for a current state we backup